### PR TITLE
FERIEIND-21 and FERIEIND-22 - Substitutes.

### PR DESCRIPTION
**Pull request by jag-itminds**
Added a Base class for SubstistutesController and ApproveReportSettingsController that Drive and Vacation can inherit from. The concrete classes then sets the SubstituteType property to their respective application. The base classes then insures that substitutes created in Vacation is only available in Vacation

Also moved from Modal-controllers from Drive to Core.

### CODE REVIEW

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 1]
```diff
@@ -0,0 +1,526 @@
+﻿module app.core.controllers {
+    "use strict";
+
+    import SubstituteType = app.core.models.SubstituteType;
+
+    export class BaseApproveReportsSettingsController {
+
+        private InfinitePeriod = 9999999999;
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: The correct casing is camelCasing not PascalCasing. It's a generel problem through out.

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+        PersonalApproverHelpText;
+
+        SubstituteOrgUnit = "";
+        ShowSubstituteSettings;
+
+        MySubstitutesGrid;
+        SubstituteGrid;
+        PersonalApproverGrid;
+
+        constructor(protected $scope, protected Person, protected $rootScope, protected Autocomplete, protected $modal, private substituteType: SubstituteType) {
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: This isn't minification safe

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+
+        OpenDeleteApprover(id) {
+            var this_ = this;
+            var modalInstance = this.$modal.open({
+                templateUrl: 'App/Core/Views/Modals/ConfirmDeleteApproverModal.html',
+                controller: 'ConfirmDeleteApproverModalInstanceController',
+                backdrop: 'static',
+                size: 'lg',
+                resolve: {
+                    persons() {
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: this isn't minification safe

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+                        return this_.People;
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: If a lambda function is used then this_ isn't needed (I believe)

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+
+        RefreshGrids() {
+            this.SubstituteGrid.dataSource.read();
+            this.MySubstitutesGrid.dataSource.read();
+            this.PersonalApproverGrid.dataSource.read();
+        }
+
+        OpenDeleteApprover(id) {
+            var this_ = this;
+            var modalInstance = this.$modal.open({
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: This modal is repeated quite a few times, but with different controller and template targets.
It smells a bit :)

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+                templateUrl: 'App/Core/Views/Modals/ConfirmDeleteApproverModal.html',
+                controller: 'ConfirmDeleteApproverModalInstanceController',
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: Controller is referenced as a string. Would be better if it was the actual class, makes it more refactor resilient

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+                this.RefreshGrids();
+            }, () => {
+
+            });
+        }
+
+        Substitutes = {
+            autoBind: false,
+            dataSource: {
+                type: "odata",
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: Aren't you using OData v4? Then it's "odata-v4"

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+                    read: {
+                        beforeSend(req) {
+                            req.setRequestHeader('Accept', 'application/json;odata=fullmetadata');
+                        },
+                        url: "odata/Substitutes/Service.Substitute(Type=" + this.substituteType +")?$expand=OrgUnit,Sub,Person,Leader &$filter=PersonId eq " + this.PersonId,
+                        dataType: "json",
+                        cache: false
+                    },
+                    parameterMap(options, type) {
+                        var d = kendo.data.transports.odata.parameterMap(options, type);
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: kendo.data.transports["odata-v4"].parameterMap(...)

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+
+                        delete d.$inlinecount; // <-- remove inlinecount parameter
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: inlinecount was removed in odata-v4

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+                        url: "odata/Substitutes/Service.Substitute(Type=" + this.substituteType +")?$expand=OrgUnit,Sub,Person,Leader &$filter=PersonId eq " + this.PersonId,
+                        dataType: "json",
+                        cache: false
+                    },
+                    parameterMap(options, type) {
+                        var d = kendo.data.transports.odata.parameterMap(options, type);
+
+                        delete d.$inlinecount; // <-- remove inlinecount parameter                                                        
+
+                        d.$count = true;
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: This is on by default when pagnig is enabled

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+                    },
+                    total(data) {
+                        return data['@odata.count']; // <-- The total items count is the data length, there is no .Count to unpack.
+                    }
+                }
+            },
+            sortable: true,
+            pageable: {
+                messages: {
+                    display: "{0} - {1} af {2} ", //{0} is the index of the first record on the page, {1} - index of the last record on the page, {2} is the total amount of records
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: Why not just use the danish translation?

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+                },
+                pageSizes: [5, 10, 20, 30, 40, 50, 100, 150, 200]
+            },
+            dataBound() {
+                this.expandRow(this.tbody.find("tr.k-master-row").first());
+            },
+            columns: [{
+                field: "Sub.FullName",
+                title: "Stedfortræder"
+            },
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: indent

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+                }, {
+                    field: "StartDateTimestamp",
+                    title: "Fra",
+                    template(data) {
+                        var m = moment.unix(data.StartDateTimestamp).toDate();
+                        return m.getDate() + "/" +
+                            (m.getMonth() + 1) + "/" + // +1 because getMonth is zero indexed.
+                            m.getFullYear();
+                    }
+            }, {
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: indent

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+                    template(data) {
+                        if (data.EndDateTimestamp === this.InfinitePeriod) {
+                            return "På ubestemt tid";
+                        }
+                        var m = moment.unix(data.EndDateTimestamp).toDate();
+                        return m.getDate() + "/" +
+                            (m.getMonth() + 1) + "/" + // +1 because getMonth is zero indexed.
+                            m.getFullYear();
+                    }
+            }, {
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: indent

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+                    title: "Til",
+                    field: "EndDateTimestamp",
+                    template(data) {
+                        if (data.EndDateTimestamp === this.InfinitePeriod) {
+                            return "På ubestemt tid";
+                        }
+                        var m = moment.unix(data.EndDateTimestamp).toDate();
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: moment should be injected into the controller

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+                        return m.getDate() + "/" +
+                            (m.getMonth() + 1) + "/" + // +1 because getMonth is zero indexed.
+                            m.getFullYear();
+                    },
+                }, {
+                    title: "Muligheder",
+                    template: "<a ng-click='ctrl.OpenEditApprover(${Id})'>Rediger</a> | <a ng-click='ctrl.OpenDeleteApprover(${Id})'>Slet</a>"
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: Use template literals

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts [hunk 17]
```diff
@@ -0,0 +1,526 @@
...
+                        var m = moment.unix(data.EndDateTimestamp).toDate();
+                        return m.getDate() + "/" +
+                            (m.getMonth() + 1) + "/" + // +1 because getMonth is zero indexed.
+                            m.getFullYear();
+                    }
+            }],
+            scrollable: false
+        }
+    }
+}
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: Missing new line at EOF

File: Presentation.Web/App/Core/Controllers/DeleteApproverModalInstanceController.js [hunk 18]
```diff
@@ -1,4 +1,4 @@
-﻿angular.module('app.drive').controller('ConfirmDeleteApproverModalInstanceController',
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/DeleteApproverModalInstanceController.js]: generated javascript files shouldn't be checked into git

File: Presentation.Web/test-results.testxml [hunk 19]
```diff
@@ -1,2 +0,0 @@
-<?xml version="1.0"?>
```
Comment (snebjorn) [inline Presentation.Web/test-results.testxml]: test results shouldn't be checked into git

File: Presentation.Web/Controllers/SubstitutesController.cs [hunk 20]
```diff
@@ -193,10 +193,10 @@ public IHttpActionResult Personal()
         /// <returns></returns>
         [EnableQuery]
         [HttpGet]
-        public IHttpActionResult Substitute()
+        public IHttpActionResult Substitute(int Type = 0)
```
Comment (snebjorn) [inline Presentation.Web/Controllers/SubstitutesController.cs]: incorrect casing

File: Presentation.Web/Controllers/SubstitutesController.cs [hunk 21]
```diff
@@ -178,10 +178,10 @@ public new IHttpActionResult Delete([FromODataUri] int key)
         /// <returns></returns>
         [EnableQuery]
         [HttpGet]
-        public IHttpActionResult Personal()
+        public IHttpActionResult Personal(int Type = 0)
```
Comment (snebjorn) [inline Presentation.Web/Controllers/SubstitutesController.cs]: incorrect casing

File: Presentation.Web/App/Vacation/app.vacation.html [hunk 22]
```diff
@@ -57,7 +57,7 @@
                     <li ui-sref-active="active" class="mainMenuLink">
                         <a ui-sref="vacation.settings"><i class="fa fa-fw fa-cogs"></i></a>
                     </li>
-                    <li ui-sref-active="active" class="ng-hide mainMenuLink" ng-hide="!showAdministration">
+                    <li ui-sref-active="active" class="ng-hide mainMenuLink">
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.html]: The ng-hide class isn't intended to be used directly. Instead use ngHide.

File: Presentation.Web/App/Vacation/Admin/AdminView.html [hunk 23]
```diff
@@ -0,0 +1,32 @@
+﻿<div class="container-fluid">
+    <div class="row">
+        <div class="col-lg-12">
+            <h1 class="page-header">
+                Administrator
+            </h1>
+            <tabset class="nav nav-pills" justified="false">
+
+                <!--<tab heading="Emailadviseringer" ng-controller="Vacation.EmailNotificationController" ng-click="amc.EmailClicked()">
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/Admin/AdminView.html]: removed unused code

File: Presentation.Web/App/Vacation/Admin/AdminMenuController.ts [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+        ];
+
+        private _currentUser;
+
+        constructor(private $scope, private Person, private $rootScope, private HelpText) {
+
+            this._currentUser = $scope.CurrentUser;
+        }
+
+        OrgSettingsClicked() {
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/Admin/AdminMenuController.ts]: camelCase

File: Presentation.Web/App/Drive/MyReports/MyRejectedReportsController.js [hunk 25]
```diff
@@ -175,7 +175,7 @@ angular.module("app.drive").controller("MyRejectedReportsController", [
            /// </summary>
            // Set initial values for kendo datepickers.
            var from = new Date();
-           from.setMonth(from.getMonth() - 3);
+           from.setMonth(from.getMonth() - 12);
```
Comment (snebjorn) [inline Presentation.Web/App/Drive/MyReports/MyRejectedReportsController.js]: Isn't this the previous year? Just -1 on the year.

File: Presentation.Web/App/Drive/MyReports/MyRejectedReportsController.js [hunk 26]
```diff
@@ -19,7 +19,7 @@ angular.module("app.drive").controller("MyRejectedReportsController", [
 
        // dates for kendo filter.
        var fromDateFilter = new Date();
-       fromDateFilter.setMonth(fromDateFilter.getMonth() - 3);
+       fromDateFilter.setMonth(fromDateFilter.getMonth() - 12);
```
Comment (snebjorn) [inline Presentation.Web/App/Drive/MyReports/MyRejectedReportsController.js]: Isn't this the previous year? Just -1 on the year.

File: Presentation.Web/App/Drive/MyReports/MyPendingReportsController.js [hunk 27]
```diff
@@ -181,7 +181,7 @@
            /// </summary>
            // Set initial values for kendo datepickers.
            var from = new Date();
-           from.setDate(from.getDate() - 365);
+           from.setDate(from.getDate() - (365*2));
```
Comment (snebjorn) [inline Presentation.Web/App/Drive/MyReports/MyPendingReportsController.js]: Magic string. Is this supposed to subtract 2 years? What happens when a year doesn't have 365 days?
Use moment instead.

File: Presentation.Web/App/Drive/MyReports/MyPendingReportsController.js [hunk 28]
```diff
@@ -18,7 +18,7 @@
 
        // dates for kendo filter.
        var fromDateFilter = new Date();
-       fromDateFilter.setDate(fromDateFilter.getDate() - 365);
+       fromDateFilter.setDate(fromDateFilter.getDate() - (365*2));
```
Comment (snebjorn) [inline Presentation.Web/App/Drive/MyReports/MyPendingReportsController.js]: Magic string. Is this supposed to subtract 2 years? What happens when a year doesn't have 365 days?
Use moment instead.

File: Presentation.Web/App/Core/Controllers/BaseSubstitutesController.ts [hunk 29]
```diff
@@ -0,0 +1,338 @@
...
+
+            });
+        }
+
+        OpenEditApprover(id) {
+            var this_ = this;
+            this.$modal.open({
+                templateUrl: "App/Core/Views/Modals/EditApproverModal.html",
+                controller: "EditApproverModalInstanceController",
+                controllerAs: "ctrl",
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Controllers/BaseSubstitutesController.ts]: Don't use ctrl as your controllerAs prefix, you lose the exact advantage you gained by using controllerAs.

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts (hunk 2 above)
Comment (jag-itminds) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: This entire class is abstract, so the concrete classes handles the injection, right? The class isn't stated as abstract, I'll have that corrected.

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts (hunk 1 above)
Comment (jag-itminds) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: It should also be noted that this entire class is just a translation from JavaScript to Typescript, no refactoring whatsoever has been done. It is the old controller ApproveReportSettingsController.js from the Drive module that has been made translated.

File: Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts (hunk 4 above)
Comment (jag-itminds) [inline Presentation.Web/App/Core/Controllers/BaseApproveReportSettingsController.ts]: I got Typescript errors using this, so I had to do this 'hack'. I suspect something is wrong elsewhere.

File: Presentation.Web/App/Drive/MyReports/MyAcceptedReportsController.js [hunk 30]
```diff
@@ -45,7 +45,7 @@ angular.module("app.drive").controller("MyAcceptedReportsController", [
                    parameterMap: function (options, type) {
                        var d = kendo.data.transports.odata.parameterMap(options);
```
Comment (snebjorn) [inline Presentation.Web/App/Drive/MyReports/MyAcceptedReportsController.js]: odata-v4

File: Presentation.Web/App/Vacation/app.vacation.routes.ts [hunk 31]
```diff
@@ -50,7 +50,27 @@
...
+                                        $rootScope.CurrentUser = res;
+                                    });
+                                } else {
+                                    return $rootScope.CurrentUser;
+                                }
+
+                            }
+                        ],
+                    }
+                });;
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.routes.ts]: double ;

File: Presentation.Web/Presentation.Web.csproj [hunk 32]
```diff
@@ -22,7 +22,7 @@
     <IISExpressAnonymousAuthentication>disabled</IISExpressAnonymousAuthentication>
     <IISExpressWindowsAuthentication>enabled</IISExpressWindowsAuthentication>
     <IISExpressUseClassicPipelineMode>false</IISExpressUseClassicPipelineMode>
-    <TypeScriptToolsVersion>1.4</TypeScriptToolsVersion>
+    <TypeScriptToolsVersion>1.7</TypeScriptToolsVersion>
```
Comment (snebjorn) [inline Presentation.Web/Presentation.Web.csproj]: newest version is 1.8.6.0

**Pull request merged**